File: PlanszaA.java
import java.util.Scanner;

public class PlanszaA {
	
	public static final int MOD=1000000103;
	
	public static boolean valid(int x){
		return x>=0;
	}
	
	public static void policz(int x, int y, int k, int[][] w, int[][] l, int[][] g, int[][] u)
	{
		if(x+y==0)return;
		if(valid(x-1)){
			g[x][y]=(g[x-1][y]+w[x-1][y])%MOD;
			if(valid(x-k-1))
				g[x][y]=(g[x][y]-w[x-k-1][y]+MOD)%MOD;
		}
		if(valid(y-1)){
			l[x][y]=(l[x][y-1]+w[x][y-1])%MOD;
			if(valid(y-k-1))
				l[x][y]=(l[x][y]-w[x][y-k-1]+MOD)%MOD;
		}
		if(valid(x-1) && valid(y-1)){
			u[x][y]=(u[x-1][y-1]+w[x-1][y-1])%MOD;
			if(valid(x-k-1) && valid(y-k-1))
				u[x][y]=(u[x][y]-w[x-k-1][y-k-1]+MOD)%MOD;
		}
		w[x][y]=((l[x][y]+g[x][y])%MOD+u[x][y])%MOD;
	}
	
	public static int solve(int n, int m, int k){
		int[][] w = new int[n][];
		int[][] l = new int[n][];
		int[][] g = new int[n][];
		int[][] u = new int[n][];
		
		for(int i=0;i<n;i++){
			w[i] = new int[m];
			l[i] = new int[m];
			g[i] = new int[m];
			u[i] = new int[m];
		}
		w[0][0]=1;
		l[0][0]=g[0][0]=u[0][0]=0;
		for(int i=0;i<n;i++)
			for(int j=0;j<m;j++)
				policz(i, j, k, w, l, g, u);
		return w[n-1][m-1];
	}
	
	public static void main(String[] args){
		int t, n, m, k;
		Scanner in = new Scanner(System.in);
		t = in.nextInt();
		for(int i=0;i<t;i++){
			n = in.nextInt();
			m = in.nextInt();
			k = in.nextInt();
			System.out.println(solve(n, m, k));
		}
	}
}

File: planszaA.c
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000103

int valid(int x){
	return x>=0;
}

void policz(int x, int y, int k, int **w, int **l, int **g, int **u){
	if(x+y==0)return;
	if(valid(x-1)){
		g[x][y]=(g[x-1][y]+w[x-1][y])%MOD;
		if(valid(x-k-1))
			g[x][y]=(g[x][y]-w[x-k-1][y]+MOD)%MOD;
	}
	if(valid(y-1)){
		l[x][y]=(l[x][y-1]+w[x][y-1])%MOD;
		if(valid(y-k-1))
			l[x][y]=(l[x][y]-w[x][y-k-1]+MOD)%MOD;
	}
	if(valid(x-1) && valid(y-1)){
		u[x][y]=(u[x-1][y-1]+w[x-1][y-1])%MOD;
		if(valid(x-k-1) && valid(y-k-1))
			u[x][y]=(u[x][y]-w[x-k-1][y-k-1]+MOD)%MOD;
	}
	w[x][y]=((l[x][y]+g[x][y])%MOD+u[x][y])%MOD;
}

int solve(int n, int m, int k){
	int i, j;
	int **w, **l, **g, **u;
	w=calloc(n, sizeof(int*));
	l=calloc(n, sizeof(int*));
	g=calloc(n, sizeof(int*));
	u=calloc(n, sizeof(int*));
	for(i=0;i<n;i++){
		w[i]=calloc(m, sizeof(int));
		l[i]=calloc(m, sizeof(int));
		g[i]=calloc(m, sizeof(int));	
		u[i]=calloc(m, sizeof(int));
	}
	w[0][0]=1;
	l[0][0]=g[0][0]=u[0][0]=0;
	for(i=0;i<n;i++)
		for(j=0;j<m;j++)
			policz(i, j, k, w, l, g, u);
	int res=w[n-1][m-1];
	for(i=0;i<n;i++){
		free(w[i]);
		free(l[i]);
		free(g[i]);
		free(u[i]);
	}
	free(w);
	free(l);
	free(g);
	free(u);
	return res;
}

int main(){
	int t, n, m, k;
	scanf("%d",&t);
	while(t--){
		scanf("%d%d%d",&n,&m,&k);
		printf("%d\n",solve(n,m,k));
	}
	return 0;
}
